README
			 GOCR (JOCR at SF.net)

GOCR is an optical character recognition program, released under the
GNU General Public License. It reads images in many formats  and outputs
a text file. Possible image formats are pnm, pbm, pgm, ppm, some pcx and
tga image files. Other formats like pnm.gz, pnm.bz2, png, jpg, tiff, gif,
bmp will be automatically converted using the netpbm-progs, gzip and bzip2
via unix pipe.
A simple graphical frontend written in tcl/tk and some
sample files (you need transfig for the sample files) are included.
Gocr is also able to recognize and translate barcodes.
You do not have to train the program or store large font bases.
Simply call gocr from the command line and get your results.

To see installation instructions, see the INSTALL file.

How to start? (QUICK START)
---------------------------
You'll probably want to use one of the frontends available, such as the TCL/TK
or the GTK. They make your life much easier.

Some examples of how you can use gocr:

  gocr -h       			# help
  gocr file.pbm				# minimum options
  gocr -v 1 file.pbm >out.txt 2>out.log # generate text- and log file
  djpeg -pnm -gray text.jpg | gocr -	# using JPEG-files
  gzip -cd text.pbm.gz | gocr -		# using gzipped PBM-files
  giftopnm text.gif | gocr -		# using GIF-files
  gocr -v 1 -v 32 -m 4 file.pbm   	# zoning and out30.bmp output
  xli -geometry 400x400 out30.bmp # see details using xli (recommanded viewer)
  wish gocr.tcl			  # X11-tcl/tk-frontend (development version)


How to get image files?
-----------------------
Scan text pages and save it as PGM/PBM/PNM file. Use a program such as
The GIMP or Sane. You can also use netpbm-progs to convert several image 
formats into PGM/PBM/PNM. The tool djpeg can be used to convert jpeg into pgm.
If you have a POSIX compatible system like linux and PNM-tools, gzip and bzip2
are installed, you are lucky and gocr will do conversion 
from [.pnm.gz, .pnm.bz2, .jpg, .jpeg, .bmp, .tiff, .png, .ps, .eps]
to [.pgm] for you. This list can easily be extended editing src/pnm.c.

Gocr also comes with some examples, try: make examples.

Memory limitations
------------------
WARNING!!!

If you use a 300dpi scan of A4 letter, the image is about 2500x3500 pixels and
gocr requires 8.75MB for storing the picture into the memory. Not only that,
but gocr may create a 2nd copy, using a total of 17MB. This is independent
of using b/w or gray-scale images. Be sure that you have enough RAM installed
in your machine! Alternatively you can cut the picture into small pieces.
You can use the pnmcut, from the netpbm package to cut the file. Example:

pnmcut -left 0 -right 2500 -top 0 -height 1000 bigfile.pnm > smallfile.pnm

And then use gocr in the cropped image as usual. Take care: if you chopp the
characters, gocr won't be able to understand that line.

Future versions will take care of this issue automatically.

Limitations
-----------
gocr is still in its early stages. Your images should fit in these requirements
if you want a good output:

- good scans (all chars well seperated, one column, no tables etc, 12pt 300dpi)
  should work well
- fonts 20-60 pixels ( 5pt * 1in/72pt * 300 dpi = 20 dots )
- output of image file for controlling detection

And note that speed is very slow (this will be changed when recognition works 
well)
  12pt 300dpi 1700x950 16lines 700chars 22x28 P90=40s..90s v0.2.3 (gcc -O0)

You can try to optimize the results:
- make good scans/treat image
- try to change the critical gray level (option -l <n>)
- control the result on out30.bmp (option -v 32)
- enlarge option -d <n> for high resolution images which are noisy
- try different combinations for option -m <n>
- for thousends of documents with same font
  you can use/create a database (-m 2/-m 130)
- use options -d 0 -m 8 on screen shots (font8x12)

What does >> NOT << work at the moment:
- complex layouts (try option -m 4)
- bad scans, noisy/snowy images, FAX-quality images
- serif fonts, italic fonts, slanted fonts
- handwritten texts (this is valid for the next ten years I guess)
   the exisctence of autotrace can shorten this 
- rotated images (but slightly rotated images should be no problem)
- small fonts (fax like) or mix of different font size
- colored images (use gray or black/white)
- Chinese, Arabian, Egyptian, Cyrillic or Klingon fonts

How it works or how it should work?
- put the entire file into RAM (300dpi grayscale recommended)
- remove dust and snow
- detect small angle (lines which are not horizontal)
- detect text boxes (option -m 4)
- detect text-lines
- detect characters
- first step recognition (every character has its own empirical procedure)
  - no neural network or similar general algorithms
- analyze not detected chars by comparison with detected ones
- try to divide overlapping chars
- testwise: compare all letters (like compression of pictures)
- for more details look to the gocr.html documentation


How can you help gocr?
----------------------
- Send comments, ideas and patches (diff -ru gocr_original/ gocr_changed/).
- If you have a lot of money, spend a bit (www.paypal.com).
- I always need example files (.pbm.gz or jpeg <100kB) for testing
  the behavior of the ocr engine under different conditions, 
  because scanning does take a lot of time which I do not have.
  But do not send files which are not convertable by commercial ocr programs
  or which are protected from copying and electronic processing by copyright.
  That will help, to get the world's best OCR open source program. :) Thanks!
- Send me your results (errors,num_chars,dpi) and if possible results
  and name of professional OCR programs for statistics.
- Read OCR literature, extract the essentials and send a short report
  to me ;).
- If you have a good idea, how to manage some OCR-tasks, tell me!
- Tell your friends about gocr. Tell me about your success. Be happy.


After all, is it gocr or jocr?
------------------------------
The original name of this project is gocr, from GNU Optical Character
Recognition. Another project is using the same name, however; so the
name was changed to jocr. If you have a good idea for a name, please
send it.


Latest news
------------
  http://jocr.sourceforge.net

Authors: (see AUTHORS)


HISTORY
History: (Changes,ChangeLog)

 0.40
       update PNM file reader to maxval > 255
       (make rpm) updated
       barcode-patch UPC_addon by Michael van Rooyen
       CAPITAL_LETTER_A_WITH_OGONEK added
       no "(PICTURE)" output for UTF8+ASCII (better for Mobile OCR project)
       smooth_borders() bug fixed and reworked
       5x7 and prop10 font adaptions
       objects now detected by flood-fill algorithm (better?)
       XML-output changed
       changed auto dust detection (not final)

 0.39
       XML output added (subject of change, suggestions are welcome)
       netpbm-link-error fixed in gocr.c and configure.in:
         gocr.c: <config.h> changed to "config.h"
       configure-option --with-netpbm=PATH and --without-netpbm added
       update configure.in according to autoconf 2.57  
       wchar_h miss-configuration fixed in pgm2asc.c
       fix compiler warnings
       char filter accepts abbreviations now, like "0-9A-F" (but slow)
       update READMEde.txt
       output barcode tags (also improved recognition)
       fix pnm.c for files like example.eps.pbm
       fix detect.c for barcodes
       fix ocr0n.c 0<->8g

 0.38
       move UTF/HTML/TeX decoding to getTextLine, return (char *) now
       out_format HTML step towards detailed XML output
       correct line detection for footnotes (detect.c)
       "y" now seen as vowel (pgm2asc.c), I<vowel> susbtituted by l<vowel>
       &eacute;-detection, &aacute;-output fixed
       default dust_size is -1 now (auto detection = mean_size/10)
       char filter added
        ex: -C 0123456789ABCDEF    - recognize only hexcodes
       man page updated (hopefully correct syntax)
       database bug fixed (small fonts, example by Chris)
       several bugs fixed by W. Webber (thanks)
       speed improved by 3rd-pass matrix filter in pixel() (pixel.c) (code from W. Webber)
       bug in remove_dust (remove.c) fixed
       for fonts bigger than 20x40 smooth_borders() changed (b/w-scans)
       bug in O0-detection fixed

 0.37
        best-fit generates probability, not perfect but better results
        bug in line detection removed (happens for lot of small boxes)
        progress output (option -x <fileID|fname>)
        counting versions number as floating point now
        MACRON and DOT_ABOVE (not complete) defined (latin2)
        adaptions for 5x7 and 6x12 screen font
	doc/ocr.tex changed to doc/gocr.html (now independent of LaTeX)
        symbols {} added
        OCR-B font tested succesfull
        better headline/picture distinction
        bug removed (struct box.modifier is wchar_t now)

        known bugs: to much newlines

 0.3.6
        CARON and Omega defined,
        output of not defined chars (HTML="&#xxx;", TeX="\symbol{xxx}")
        system dependend bug: isupper(>255) SIGSEGV fixed
        better line detection for lines with lowercase chars only
        lot of possible SIGSEGV in list_del() fixed
        barcode recognition (UPC,code128)
        .ps .eps via pstopnm supported
        -m 256 switches off the main ocr engine (usefull together with -m 2 for identical chars)
        strings added to database ("ff","ft","special-symbol")
        gocr.tcl adapted to gocr v0.3
        internal detection probability introduced

 0.3.5
        minor and major fixes (string\0 bugs)
        memory leak fixes by Duncan Edwards
        layout analysis or zoning (-m 4) improved, 
        now it detects pictures and columns much better
        the behavior of setting threshold (-l) is slightly changed
        wcsdup defined for non-gnu-systems (BSD), further Problems?
        better context correction for 10 (IO,lO)
        Fixes for S.Koledin examples "GlS"
        Euro-currency-sign detection added
        better pitch estimation for proportional font (needs to be improved)
        make install DESTDIR= instead configure --prefix= (better?)
        use wchar_t by default, more simple code and -f works with nonLinuxOS
        line detection more robust against vertical glued chars (js)
        -f UTF8 added (usefull for xterm -u8), should be default?
        handle vertical glued boxes (ex: g over T)
 0.3.4
        some BSD adaptions (no WCHAR?), tell me if there are still problems
        use unicode in database (4-8 hex digits)
        new option: -p database_path/
        TILDE fixed, #, &AElig;, &Aring;, etc. added (swedish,norwegian)
        layout analysis improved
 0.3.3
        database (-m 2) bug fixed and interactive mode (-m 130) added
          its not finished, but you can test it
          result should be ok for machine generated images (no scans)           
        engine improved a bit
 0.3.2
        ocr-engine improved for screen fonts (thanks for examples)
        option -f [HTML,TeX,...] added
 0.3.1
        make install updated
 0.3.0  some parts of the code reviewed (most work done by Bruno Barberi Gnecco)
        tkispell patch from David Pinson (exec bug fixed)
	gnome frontend added (Dany De Bontrider)
        acute, grave, circumflex ... detection
        C++ parts rewritten into C, and much more (see REVIEW)
 0.2.7  lib-patch from Klaas Freitag inserted, engine improved
        option -n 1 detect only numbers, get threshold value by otsu.cc
        xxx.pnm.bz2 can be used on linux systems bzip2 installed
 0.2.6  pipes used on POSIX2-systems for easier use of jpg,gif,tiff,pnm.gz-files
        example: gocr text.jpg; gocr text.pnm.gz
        verbose output on stderr, text output on stdout,
        redirection of output possible (-e, -o, example: -e /dev/stdout)
        engine upgraded a bit  (thx for the new sample files)
        gocr.tcl upgraded (save options, save text)
        DOS/WIN95-EXE created, download GOCREXE.ZIP (v0.2.5)
 0.2.5  program convert renamed to jconv
        you can choose stdin as input now, for using conversion tools
	example: djpeg -pnm -gray text.jpg | gocr -i -
        option "--help" added, some bugs removed

src/pnm.c
/* (c) Joerg Schulenburg 2000
    v0.1.0 initial version (stdin added)
    v0.2.0 popen added
    v0.2.7 review by Bruno Barberi Gnecco
    v0.39  autoconf
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <unistd.h>

#include "pnm.h"
#ifdef HAVE_PAM_H
# include <pam.h>
# include <sys/types.h>
# include <sys/stat.h>
# include <fcntl.h>
#else
# include <ctype.h>
#endif

#define EE()         fprintf(stderr,"\nERROR "__FILE__" L%d: ",__LINE__)
#define E0(x0)       {EE();fprintf(stderr,x0 "\n");      }
#define F0(x0)       {EE();fprintf(stderr,x0 "\n");      exit(1);}
#define F1(x0,x1)    {EE();fprintf(stderr,x0 "\n",x1);   exit(1);}
#define F2(x0,x1,x2) {EE();fprintf(stderr,x0 "\n",x1,x2);exit(1);}

/*
 * Weights to use for the different colours when converting a ppm
 * to greyscale.  These weights should sum to 1.0
 * 
 * The below values have been chosen to reflect the fact that paper
 * goes a reddish-yellow as it ages.
 */
#define PPM_RED_WEIGHT .499
#define PPM_GREEN_WEIGHT .387
#define PPM_BLUE_WEIGHT .114

/*
    feel free to expand this list of usable converting programs
    Note 1: the last field must be NULL.
    Note 2: "smaller" extensions must come later: ".pnm.gz" must come
       before ".pnm".
 */
char *xlist[]={
  ".pnm.gz",	"gzip -cd",  /* compressed pnm-files, gzip package */
  ".pbm.gz",	"gzip -cd",
  ".pgm.gz",	"gzip -cd",
  ".ppm.gz",	"gzip -cd",
  ".pnm.bz2",	"bzip2 -cd",
  ".pbm.bz2",	"bzip2 -cd",
  ".pgm.bz2",	"bzip2 -cd",
  ".ppm.bz2",	"bzip2 -cd",
  ".jpg", 	"djpeg -gray -pnm",  /* JPG/JPEG, jpeg package */
  ".jpeg",	"djpeg -gray -pnm",
  ".gif",	"giftopnm",          /* GIF, netpbm package */
  ".bmp",	"bmptoppm",
  ".tiff",	"tifftopnm",
  ".png",	"pngtopnm", /* Portable Network Graphics (PNG) format */
  ".ps",	"pstopnm -stdout -portrait -pgm", /* postscript */
  ".eps",	"pstopnm -stdout -portrait -pgm", /* encapsulated postscript */
  ".fig",	"fig2dev -L ppm -m 3", /* xfig files, transfig package */
  NULL
};

/* return a pointer to command converting file to pnm or NULL */
char *testsuffix(char *name){
  int i; char *rr;

  for(i = 0; xlist[i] != NULL; i += 2 ) {
    if((rr=strstr(name, xlist[i])) != NULL)
      if(strlen(rr)==strlen(xlist[i])) /* handle *.eps.pbm correct */
        return xlist[i+1];
  }
  return NULL;
}


char read_char(FILE *f1){	// filter #-comments
  char c;
  int  m;
  for(m=0;;){
    c=fgetc(f1);
    if( feof(f1)   )		E0("read feof");
    if( ferror(f1) )		F0("read ferror");
    if( c == '#'  )		{ m = 1; continue; }
    if( m ==  0   )		return c;
    if( c == '\n' )		m = 0;
  }
}

/*
   for simplicity only PAM of netpbm is used, the older formats
   PBM, PGM and PPM can be handled implicitly by PAM routines (js05)
 */
#ifdef HAVE_PAM_H
void readpgm(char *name, pix * p, int vvv) {
  FILE *fp;
  char *pip, magic1, magic2;
  int i, j, sample, minv = 0, maxv = 0;
  int ppm_format = 0;
  struct pam inpam;
  tuple *tuplerow;

  assert(p);

  /* open file; test if conversion is needed. */
  if (name[0] == '-' && name[1] == '\0') {
    fp = stdin;
  }
  else {
    pip = testsuffix(name);
    if (!pip) {
      fp = fopen(name, "rb");
      if (!fp)
	F1("opening file %s", name);
    }
    else {
      char *buf = (char *)malloc((strlen(pip)+strlen(name)+2)*
	  sizeof(char));
      sprintf(buf, "%s %s", pip, name);
      if (vvv) {
	fprintf(stderr, "# popen( %s )\n", buf);
      }
#ifdef HAVE_POPEN
      fp = popen(buf, "r");
#else
      F0("sorry, compile with HAVE_POPEN to use pipes");
#endif
      if (!fp)
	F1("opening pipe %s", buf);
      free(buf);
    }
  }

  /* read pgm */
  pnm_readpaminit(fp, &inpam, sizeof(inpam));

  p->x = inpam.width;
  p->y = inpam.height;
  if ( !(p->p = (unsigned char *)malloc(p->x*p->y)) )
    F1("Error at malloc: p->p: %d bytes", p->x*p->y);
  tuplerow = pnm_allocpamrow(&inpam);
  magic1=(inpam.format >> 8) & 255; /* 'P' for PNM,PAM */
  magic2=(inpam.format     ) & 255; /* '7' for PAM */
  minv=inpam.maxval;
  if (vvv) {
      fprintf(stderr, "# readpam: format=0x%04x=%c%c h*w(d*b)=%d*%d(%d*%d)\n",
        inpam.format, /* magic1*256+magic2 */
        ((magic1>31 && magic1<127)?magic1:'.'),
        ((magic2>31 && magic2<127)?magic2:'.'),
        inpam.height,
        inpam.width,
        inpam.depth,
        inpam.bytes_per_sample);
  }
  for ( i=0; i < inpam.height; i++ ) {
    pnm_readpamrow(&inpam, tuplerow);
    for ( j = 0; j < inpam.width; j++ ) {
      if (inpam.depth>=3)
        sample
          = PPM_RED_WEIGHT   * tuplerow[j][0]
          + PPM_GREEN_WEIGHT * tuplerow[j][1]
          + PPM_BLUE_WEIGHT  * tuplerow[j][2];
      else
        sample = tuplerow[j][0];
      sample = 255 * sample / inpam.maxval;
      p->p[i*inpam.width+j] = sample;
      if (maxv<sample) maxv=sample;
      if (minv>sample) minv=sample;
    }
  }
  if (vvv) fprintf(stderr,"# readpam: min=%d max=%d\n",minv,maxv);
  pnm_freepamrow(tuplerow);
  fclose(fp);

  p->bpp = 1;
}

#else
/*
   if PAM not installed, here is the fallback routine,
   which is not so powerful
 */
void readpgm(char *name,pix *p,int vvv){	// raw-ppm
  char c1,c2,*pip;     int nx,ny,nc,mod,i,j,k;	// buffer
  FILE *f1;            unsigned char *pic;
  char buf[512];
  int lx, ly, dx;
  int value;
  int bps=1; /* bytes per sample (0..255..65535...) */

  pip=NULL;
  if(name[0]=='-' && name[1]==0){
    f1=stdin;  /* is this correct ??? */
  } else {
    pip=testsuffix(name);
    if(!pip){
      f1=fopen(name,"rb"); if(!f1)F1("opening file %s",name);
    } else {
      sprintf(buf,"%s %s",pip,name); /* how to prevent OVL ? */
      if(vvv){ fprintf(stderr,"# popen( %s )\n",buf); }
#ifdef HAVE_POPEN
      f1=popen(buf,"r");
#else
      F0("sorry, compile with HAVE_POPEN to use pipes");
#endif
      if(!f1)F1("opening pipe %s",buf);
    }
  }
  c1=fgetc(f1); if(feof(f1))F0("read");
  c2=fgetc(f1); if(feof(f1))F0("read");
  if(c1!='P')F0("first byte should be P in PNM-files");
  if(c2!='5' && c2!='4' && c2!='6')// F0("no PNM-RAW"); // PGM-RAW,PBM-RAW,PPM-RAW
  if(c2!='1' && c2!='2' && c2!='3')F0("no PNM");
  nx=ny=nc=0; if( c2=='4' || c2=='1') nc=1;
  for(mod=0;((c2=='5' || c2=='2') && (mod&7)<6)
        ||  ((c2=='6' || c2=='3') && (mod&7)<6)		
        ||  ((c2=='4' || c2=='1') && (mod&7)<4);)		
  {						// mode: 0,2,4=[ |\t|\r|\n] 
  						//   1=nx 3=ny 5=nc 8-13=#rem
    c1=read_char(f1);				// former: # mod|=8
    if( (mod & 1)==0 )				// whitespaces
    if( !isspace(c1) ) mod++;
    if( (mod & 1)==1 ) {
      if( !isdigit(c1) ) {
        if( !isspace(c1) )F0("unexpected character");
        mod++; }
      else if(mod==1) nx=nx*10+c1-'0';
      else if(mod==3) ny=ny*10+c1-'0';
      else if(mod==5) nc=nc*10+c1-'0';
    }
  }
  if(vvv)
   fprintf(stderr,"# PNM P%c x=%d y=%d c=%d head=%ld",c2,nx,ny,nc,ftell(f1));
  if( c2=='4' && (nx&7)!=0 ){
    /* nx=(nx+7)&~7;*/ if(vvv)fprintf(stderr," PBM2PGM nx %d",(nx+7)&~7);
  }
  if (nc>> 8) bps=2;
  if (nc>>16) bps=3;
  if (nc>>24) bps=4;
  fflush(stdout);
  pic=(unsigned char *)malloc( nx*ny );
  if(pic==NULL)F0("memory failed");			// no memory
  /* this is a slow but short routine for P1 to P6 formats */
  if( c2=='5' && nc<256 ) /* fast */
    if(ny!=(int)fread(pic,   nx,ny,f1))F0("read");	// read all lines
  if( c2=='5' && nc>255 ) /* slow */
    for (i=0;i<nx*ny;i++) {
     if(bps!=(int)fread(buf,1,bps,f1))F0("read");	// read one value
     pic[i]=buf[bps-1]; /* store the most significant byte */
    }
  if( c2=='2' )for(mod=j=i=0;i<nx*ny && !feof(f1);){ // not very clever
    c1=read_char(f1);
    if( !isdigit(c1) ) {  if( !isspace(c1) )F0("unexpected char");
      if(1&mod) { pic[i]=j*255/nc; i++; j=0; mod=0; }  }
    else { j=j*10+c1-'0'; mod=1; };
  }
  if( c2=='1' )for(mod=j=i=0,nc=255;i<nx*ny && !feof(f1);){ // PBM-ASCII 0001100
    c1=read_char(f1);
    if( isdigit(c1) ) { pic[i]=((c1=='0')?255:0); i++; }
    else if( !isspace(c1) )F0("unexpected char");
  }
  if( c2=='3' )for(mod=k=j=i=0;i<nx*ny*3 && !feof(f1);){
    c1=read_char(f1);
    if( !isdigit(c1) ) {  if( !isspace(c1) )F0("unexpected char");
      if(1&mod) { k+=j; if(mod==5){ pic[i]=k/3; i++; }
        j=0; mod=(mod+1)%6; }  }
    else { j=j*10+c1-'0'; if(!(mod&1)) mod++; };
  }
  if( c2=='6' ){
    for (i=0;i<nx*ny;i++) {
      if(3*bps!=(int)fread(buf,1,3*bps,f1))F0("read");	// read all lines
      pic[i]=( buf[  bps-1]
              +buf[2*bps-1]
              +buf[3*bps-1] ) / 3;
    }
  }
  if( c2=='4' ){				// PBM-RAW
    dx=(nx+7)&~7;				// dx (mod 8)
    if(ny!=(int)fread(pic,dx>>3,ny,f1))F0("read");	// read all bytes
    for(ly=ny-1;ly>=0;ly--)
    for(lx=nx-1;lx>=0;lx--)
    pic[lx+ly*nx]=( (128 & pic[(lx+ly*dx)>>3]<<(lx & 7)) ? 0 : 255 );
    nc=255;
  }
  if(name[0]=='-' && name[1]==0){
    f1=NULL;
  } else {
    if(!pip) fclose(f1);
#ifdef HAVE_POPEN
    else     pclose(f1);	/* close pipe (Jul00) */
#endif
  }
  p->p=pic;  p->x=nx;  p->y=ny; p->bpp=1;
  if(vvv)fprintf(stderr,"\n");
}
#endif /* HAVE_PAM_H */

int writepgm(char *nam,pix *p){// P5 raw-pgm
  FILE *f1;int a,x,y;
  f1=fopen(nam,"wb");if(!f1)F0("open");		// open-error
  fprintf(f1,"P5\n%d %d\n255\n",p->x,p->y);
  if(p->bpp==3)
  for(y=0;y<p->y;y++)
  for(x=0;x<p->x;x++){	// set bit
    a=x+y*p->x;
    p->p[a]=(p->p[3*a+0]+p->p[3*a+1]+p->p[3*a+2])/3;
  }
  if(p->y!=(int)fwrite(p->p,p->x,p->y,f1))F0("write");	// write all lines
  fclose(f1);
  return 0;
}

// pgm2ppm, use last 3 bits for farbcoding 
int writeppm(char *nam,pix *p){// P6 raw-ppm
  FILE *f1;int x,y; unsigned char rgb[3];
  f1=fopen(nam,"wb");if(!f1)F0("open");		// open-error
  fprintf(f1,"P6\n%d %d\n255\n",p->x,p->y);
  if( p->bpp==1 )
  for(y=0;y<p->y;y++)
  for(x=0;x<p->x;x++){
    rgb[0]=rgb[1]=rgb[2]=p->p[x+y*p->x];
    if(p->p[x+y*p->x]&4) rgb[0]=64;
    if(p->p[x+y*p->x]&2) rgb[1]=64;
    if(p->p[x+y*p->x]&1) rgb[2]=64;
    if( 1!=(int)fwrite(rgb,3,1,f1) )F0("write");
  }
  if( p->bpp==3 )
  if( p->y!=(int)fwrite(p->p,3*p->x,p->y,f1) )F0("write");
  fclose(f1);
  return 0;
}

// high bit = first, 
int writepbm(char *nam,pix *p){// P4 raw-pbm
  FILE *f1;int x,y,a,b,dx,i;
  dx=(p->x+7)&~7;	// enlarge to a factor of 8
  for(y=0;y<p->y;y++)
  for(x=0;x<p->x;x++){	// set bit
    a=(x+y*dx)>>3;b=7-(x&7);	// adress an bitisnumber
    i=x+y*p->x;
    if(p->bpp==3) i=(p->p[3*i+0]+p->p[3*i+1]+p->p[3*i+2])/3;
    else         i= p->p[  i  ];
    i=((i>127)?0:1);
    p->p[a]=(p->p[a] & (~1<<b)) | i<<b;
  }
  f1=fopen(nam,"wb");if(!f1)F0("open");		// open-error
  fprintf(f1,"P4\n%d %d\n",p->x,p->y);
  if(p->y!=(int)fwrite(p->p,dx>>3,p->y,f1))F0("write");	// write all lines
  fclose(f1);
  return 0;
}
// ------------------------------------------------------------------------
